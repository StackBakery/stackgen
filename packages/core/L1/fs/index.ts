export * from "./File";
export * from "./JsonFile";
export * from "./YamlFile";
export * from "./TemplateFile";
