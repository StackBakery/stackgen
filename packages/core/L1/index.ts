export * from "./traits";
export * from "./fs";
export * from "./project";
export * from "./scripts";
export * from "./synthesizers";
