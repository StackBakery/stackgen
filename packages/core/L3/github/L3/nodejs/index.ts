export * from "./npm/steps";
export * from "./yarn/steps";
