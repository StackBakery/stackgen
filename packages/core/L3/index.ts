export * from "./cloud";
export * from "./github";
export * from "./nodejs";
