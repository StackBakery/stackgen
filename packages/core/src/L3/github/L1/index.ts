export * from "./GithubJob";
export * from "./GithubStep";
export * from "./GithubWorkflow";
