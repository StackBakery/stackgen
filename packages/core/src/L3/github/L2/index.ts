export * from "./jobs";
export * from "./steps";
export * from "./workflows";
export * from "./GithubSupport";
