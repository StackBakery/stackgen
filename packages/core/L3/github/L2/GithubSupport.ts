import { Construct } from "constructs";
import { Bindings, File, Project } from "../../../L1";
import { SemanticPullRequestLintWorkflowProps, PullRequestLintWorkflow } from "./index";

export interface GithubSupportProps {
  readonly pullRequestLint?: Omit<SemanticPullRequestLintWorkflowProps, "runsOn">;
  readonly pullRequestTemplate?: string;
}

export class GithubSupport extends Construct {
  constructor(scope: Construct, id: string, props?: GithubSupportProps) {
    super(scope, id);

    if (props?.pullRequestLint) {
      new PullRequestLintWorkflow(this, "pull-request-lint", props?.pullRequestLint);
    }

    if (props?.pullRequestTemplate) {
      new File(this, "PullRequestTemplate", { filePath: ".github/pull_request_template.md", content: props.pullRequestTemplate });
    }

    Bindings.of(Project.of(this)).bind(this);
  }
}
