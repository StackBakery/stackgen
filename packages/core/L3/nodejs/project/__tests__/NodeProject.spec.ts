import { FileSynthesizer, PostInstallShellScript } from "../../../../L1";
import { Manifest } from "../../../../L2";
import { NodeProject, NodeWorkspace, TypescriptSupport } from "../../../../L3";

describe("NodeProject", () => {
  it("creates a generic NodeJS project", async () => {
    const workspace = new NodeWorkspace();
    const project = new NodeProject(workspace, "MyProject", {
      author: {
        email: "test@example.com",
        name: "Test Example",
        organization: true,
      },
      description: "MyDescription",
      repository: {
        type: "git",
        url: "http://localhost",
      },
      scripts: {
        test: "./myTestScript.sh",
      },
      private: true,
      resolutions: { a: "b" },
      dependencies: ["@package/latest", { name: "@package/version", version: "^1" }],
      devDependencies: ["@devPackage/latest", { name: "@devPackage/version", version: "^2" }],
      peerDependencies: ["@peerPackage/latest", { name: "@peerPackage/version", version: "^3" }],
      bundledDependencies: ["@devPackage/latest"],
    });

    workspace.synth();

    const manifest = Manifest.of(project);
    const fields = manifest.fields;

    expect(fields.name).toBe("my-project");
    expect(fields.author).toStrictEqual({
      email: "test@example.com",
      name: "Test Example",
      organization: true,
      url: undefined,
    });
    expect(fields.description).toBe("MyDescription");
    expect(fields.private).toBe(true);
    expect(fields.main).toBe("build/index.js");
    expect(fields.files).toStrictEqual(["build/*.js", "build/**/*.js"]);
    expect(fields.dependencies).toStrictEqual({ "@package/latest": "*", "@package/version": "^1" });
    expect(fields.peerDependencies).toStrictEqual({ "@peerPackage/latest": "*", "@peerPackage/version": "^3" });
    expect(fields.devDependencies).toStrictEqual({
      "@pdkit/cli": "*",
      "@pdkit/core": "*",
      "@devPackage/latest": "*",
      "@devPackage/version": "^2",
    });
    expect(fields.bundledDependencies).toStrictEqual(["@devPackage/latest"]);
    expect(fields.resolutions).toStrictEqual({ a: "b" });
    expect(fields.repository).toStrictEqual({
      type: "git",
      url: "http://localhost",
    });
    expect(fields.jest).toBeUndefined();
    expect(fields.eslintConfig).toBeUndefined();
    expect(fields.publishConfig).toBeUndefined();
    expect(fields.scripts).toStrictEqual({
      test: "./myTestScript.sh",
    });
    expect(fields.a).toBeUndefined();
    expect(manifest.fields).toMatchSnapshot();

    const installCommand = project.node.tryFindChild("InstallCommand") as PostInstallShellScript;

    expect(installCommand).not.toBeUndefined();
    expect(installCommand.command).toStrictEqual(["npm", "install"]);
  });

  it("creates a generic NodeJS project with Yalc", async () => {
    const workspace = new NodeWorkspace();
    const project = new NodeProject(workspace, "MyProject", {
      description: "MyDescription",
      repository: {
        type: "git",
        url: "http://localhost",
      },
      private: true,
      resolutions: { a: "b" },
      yalc: true,
    });

    workspace.synth();

    const manifest = Manifest.of(project);
    const fields = manifest.fields;

    expect(fields.name).toBe("my-project");
    expect(fields.description).toBe("MyDescription");
    expect(fields.private).toBe(true);
    expect(fields.main).toBe("build/index.js");
    expect(fields.files).toStrictEqual(["build/*.js", "build/**/*.js"]);
    expect(fields.devDependencies).toStrictEqual({ "@pdkit/cli": "*", "@pdkit/core": "*" });
    expect(fields.resolutions).toStrictEqual({ a: "b" });
    expect(fields.repository).toStrictEqual({
      type: "git",
      url: "http://localhost",
    });
    expect(fields.jest).toBeUndefined();
    expect(fields.eslintConfig).toBeUndefined();
    expect(fields.publishConfig).toBeUndefined();
    expect(fields.scripts).toStrictEqual({ yalc: "npx yalc publish" });
    expect(fields.a).toBeUndefined();
    expect(manifest.fields).toMatchSnapshot();
  });

  it("creates a generic NodeJS project with overloaded packageName", async () => {
    const workspace = new NodeWorkspace();
    const project = new NodeProject(workspace, "MyProject", {
      packageName: "@example/test",
      description: "MyDescription",
      repository: {
        type: "git",
        url: "http://localhost",
      },
      private: true,
      resolutions: { a: "b" },
      yalc: true,
    });

    workspace.synth();

    const manifest = Manifest.of(project);
    const fields = manifest.fields;

    expect(fields.name).toBe("@example/test");
    expect(fields.description).toBe("MyDescription");
    expect(fields.private).toBe(true);
    expect(fields.main).toBe("build/index.js");
    expect(fields.files).toStrictEqual(["build/*.js", "build/**/*.js"]);
    expect(fields.devDependencies).toStrictEqual({ "@pdkit/cli": "*", "@pdkit/core": "*" });
    expect(fields.resolutions).toStrictEqual({ a: "b" });
    expect(fields.repository).toStrictEqual({
      type: "git",
      url: "http://localhost",
    });
    expect(fields.jest).toBeUndefined();
    expect(fields.eslintConfig).toBeUndefined();
    expect(fields.publishConfig).toBeUndefined();
    expect(fields.scripts).toStrictEqual({ yalc: "npx yalc publish" });
    expect(fields.a).toBeUndefined();
    expect(manifest.fields).toMatchSnapshot();
  });

  it("creates a generic NodeJS project with a license", async () => {
    const workspace = new NodeWorkspace();
    const project = new NodeProject(workspace, "MyProject", {
      license: "Apache-2.0",
      description: "MyDescription",
      repository: {
        type: "git",
        url: "http://localhost",
      },
      private: true,
      resolutions: { a: "b" },
      yalc: true,
    });

    workspace.synth();

    const manifest = Manifest.of(project);
    const fields = manifest.fields;
    const vfs = FileSynthesizer.of(project);

    expect(vfs.virtualFiles).toContain("LICENSE");

    expect(fields.name).toBe("my-project");
    expect(fields.description).toBe("MyDescription");
    expect(fields.private).toBe(true);
    expect(fields.license).toBe("Apache-2.0");
    expect(fields.main).toBe("build/index.js");
    expect(fields.files).toStrictEqual(["build/*.js", "build/**/*.js"]);
    expect(fields.devDependencies).toStrictEqual({ "@pdkit/cli": "*", "@pdkit/core": "*" });
    expect(fields.resolutions).toStrictEqual({ a: "b" });
    expect(fields.repository).toStrictEqual({
      type: "git",
      url: "http://localhost",
    });
    expect(fields.jest).toBeUndefined();
    expect(fields.eslintConfig).toBeUndefined();
    expect(fields.publishConfig).toBeUndefined();
    expect(fields.scripts).toStrictEqual({ yalc: "npx yalc publish" });
    expect(fields.a).toBeUndefined();
    expect(manifest.fields).toMatchSnapshot();
  });

  it("creates a generic NodeJS project with Jest", async () => {
    const workspace = new NodeWorkspace();
    const project = new NodeProject(workspace, "MyProject", {
      description: "MyDescription",
      repository: {
        type: "git",
        url: "http://localhost",
      },
      private: true,
      resolutions: { a: "b" },
      jest: {},
    });

    workspace.synth();

    const manifest = Manifest.of(project);
    const fields = manifest.fields;

    expect(fields.name).toBe("my-project");
    expect(fields.description).toBe("MyDescription");
    expect(fields.private).toBe(true);
    expect(fields.main).toBe("build/index.js");
    expect(fields.files).toStrictEqual(["build/*.js", "build/**/*.js"]);
    expect(fields.devDependencies).toStrictEqual({ "@pdkit/cli": "*", "@pdkit/core": "*", "jest": "*", "jest-junit": "^13" });
    expect(fields.resolutions).toStrictEqual({ a: "b" });
    expect(fields.repository).toStrictEqual({
      type: "git",
      url: "http://localhost",
    });
    expect(fields.jest).not.toBeUndefined();
    expect(fields.eslintConfig).toBeUndefined();
    expect(fields.publishConfig).toBeUndefined();
    expect(fields.a).toBeUndefined();
    expect(manifest.fields.jest).toMatchSnapshot();
  });

  it("creates a generic NodeJS project with Eslint", async () => {
    const workspace = new NodeWorkspace();
    const project = new NodeProject(workspace, "MyProject", {
      description: "MyDescription",
      repository: {
        type: "git",
        url: "http://localhost",
      },
      private: true,
      resolutions: { a: "b" },
      eslint: {},
    });

    workspace.synth();

    const manifest = Manifest.of(project);
    const fields = manifest.fields;

    expect(fields.name).toBe("my-project");
    expect(fields.description).toBe("MyDescription");
    expect(fields.private).toBe(true);
    expect(fields.main).toBe("build/index.js");
    expect(fields.files).toStrictEqual(["build/*.js", "build/**/*.js"]);
    expect(fields.devDependencies).toStrictEqual({
      "@pdkit/cli": "*",
      "@pdkit/core": "*",
      "eslint": "*",
      "eslint-import-resolver-alias": "*",
      "eslint-import-resolver-node": "*",
      "eslint-plugin-import": "*",
    });
    expect(fields.resolutions).toStrictEqual({ a: "b" });
    expect(fields.repository).toStrictEqual({
      type: "git",
      url: "http://localhost",
    });
    expect(fields.jest).toBeUndefined();
    expect(fields.eslintConfig).not.toBeUndefined();
    expect(fields.publishConfig).toBeUndefined();
    expect(fields.a).toBeUndefined();
    expect(manifest.fields.eslintConfig).toMatchSnapshot();
  });

  it("creates a generic NodeJS project with Typescript", async () => {
    const workspace = new NodeWorkspace();
    const project = new NodeProject(workspace, "MyProject", {
      description: "MyDescription",
      repository: {
        type: "git",
        url: "http://localhost",
      },
      private: true,
      resolutions: { a: "b" },
      typescript: {},
    });

    workspace.synth();

    const manifest = Manifest.of(project);
    const fields = manifest.fields;
    const ts = TypescriptSupport.of(project);

    expect(fields.name).toBe("my-project");
    expect(fields.description).toBe("MyDescription");
    expect(fields.private).toBe(true);
    expect(fields.main).toBe("build/index.js");
    expect(fields.files).toStrictEqual(["build/*.js", "build/**/*.js", "build/*.d.ts", "build/**/*.d.ts"]);
    expect(fields.devDependencies).toStrictEqual({
      "@pdkit/cli": "*",
      "@pdkit/core": "*",
      "@types/node": "*",
      "ts-node": "*",
      "typescript": "*",
    });
    expect(fields.resolutions).toStrictEqual({ a: "b" });
    expect(fields.repository).toStrictEqual({
      type: "git",
      url: "http://localhost",
    });
    expect(fields.jest).toBeUndefined();
    expect(fields.eslintConfig).toBeUndefined();
    expect(fields.publishConfig).toBeUndefined();
    expect(fields.a).toBeUndefined();

    const tsFields = ts.file.fields;

    expect(tsFields).toMatchSnapshot();
  });

  it("creates a generic NodeJS project with Eslint, Jest, and Typescript", async () => {
    const workspace = new NodeWorkspace();
    const project = new NodeProject(workspace, "MyProject", {
      description: "MyDescription",
      repository: {
        type: "git",
        url: "http://localhost",
      },
      private: true,
      resolutions: { a: "b" },
      typescript: {},
      jest: {},
      eslint: {},
    });

    workspace.synth();

    const manifest = Manifest.of(project);
    const fields = manifest.fields;
    const ts = TypescriptSupport.of(project);

    expect(fields.name).toBe("my-project");
    expect(fields.description).toBe("MyDescription");
    expect(fields.private).toBe(true);
    expect(fields.main).toBe("build/index.js");
    expect(fields.files).toStrictEqual(["build/*.js", "build/**/*.js", "build/*.d.ts", "build/**/*.d.ts"]);
    expect(fields.devDependencies).toStrictEqual({
      "@pdkit/cli": "*",
      "@pdkit/core": "*",
      "@types/node": "*",
      "@types/jest": "*",
      "@typescript-eslint/eslint-plugin": "*",
      "@typescript-eslint/parser": "*",
      "ts-node": "*",
      "ts-jest": "*",
      "typescript": "*",
      "eslint": "*",
      "eslint-import-resolver-alias": "*",
      "eslint-import-resolver-node": "*",
      "eslint-import-resolver-typescript": "*",
      "eslint-plugin-import": "*",
      "jest": "*",
      "jest-junit": "^13",
    });
    expect(fields.resolutions).toStrictEqual({ a: "b" });
    expect(fields.repository).toStrictEqual({
      type: "git",
      url: "http://localhost",
    });
    expect(fields.jest).not.toBeUndefined();
    expect(fields.eslintConfig).not.toBeUndefined();
    expect(fields.publishConfig).toBeUndefined();
    expect(fields.a).toBeUndefined();

    const tsFields = ts.file.fields;

    expect(manifest.fields.jest).toMatchSnapshot();
    expect(manifest.fields.eslintConfig).toMatchSnapshot();
    expect(tsFields).toMatchSnapshot();
  });

  it("supports object supplied dependencies", () => {
    const workspace = new NodeWorkspace();

    const project = new NodeProject(workspace, "MyProject", {
      dependencies: {
        a: "^7",
      },
    });

    workspace.synth();

    const manifest = Manifest.of(project);
    const fields = manifest.fields;

    expect(fields.dependencies).toStrictEqual({ a: "^7" });
  });

  it("supports object supplied dependencies with merge", () => {
    const workspace = new NodeWorkspace();

    const project = new NodeProject(workspace, "MyProject", {
      devDependencies: {
        a: "^7",
      },
    });

    workspace.synth();

    const manifest = Manifest.of(project);
    const fields = manifest.fields;

    expect(fields.devDependencies).toStrictEqual({ "a": "^7", "@pdkit/cli": "*", "@pdkit/core": "*" });
  });
});
