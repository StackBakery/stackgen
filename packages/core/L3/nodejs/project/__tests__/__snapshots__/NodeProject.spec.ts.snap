// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeProject creates a generic NodeJS project 1`] = `
{
  "author": {
    "email": "test@example.com",
    "name": "Test Example",
    "organization": true,
    "url": undefined,
  },
  "bin": undefined,
  "bugs": undefined,
  "bundledDependencies": [
    "@devPackage/latest",
  ],
  "dependencies": {
    "@package/latest": "*",
    "@package/version": "^1",
  },
  "description": "MyDescription",
  "devDependencies": {
    "@devPackage/latest": "*",
    "@devPackage/version": "^2",
    "@stackgen/cli": "*",
    "@stackgen/core": "*",
  },
  "files": [
    "build/*.js",
    "build/**/*.js",
  ],
  "homepath": undefined,
  "keywords": undefined,
  "main": "build/index.js",
  "man": undefined,
  "name": "my-project",
  "peerDependencies": {
    "@peerPackage/latest": "*",
    "@peerPackage/version": "^3",
  },
  "private": true,
  "repository": {
    "type": "git",
    "url": "http://localhost",
  },
  "resolutions": {
    "a": "b",
  },
  "scripts": {
    "test": "./myTestScript.sh",
  },
  "types": undefined,
  "version": "0.0.1",
}
`;

exports[`NodeProject creates a generic NodeJS project with Eslint 1`] = `
{
  "env": {
    "jest": false,
    "node": true,
  },
  "extends": [
    "plugin:import/recommended",
  ],
  "ignorePatterns": [
    "node_modules/",
    "*.generated.ts",
    "coverage",
  ],
  "overrides": [
    {
      "files": [
        ".stackgenrc.ts",
      ],
      "rules": {
        "@typescript-eslint/no-require-imports": "off",
        "import/no-extraneous-dependencies": "off",
      },
    },
  ],
  "parser": undefined,
  "parserOptions": {
    "ecmaVersion": 2020,
  },
  "plugins": [],
  "root": true,
  "rules": {
    "array-bracket-newline": [
      "error",
      "consistent",
    ],
    "array-bracket-spacing": [
      "error",
      "never",
    ],
    "brace-style": [
      "error",
      "1tbs",
      {
        "allowSingleLine": true,
      },
    ],
    "comma-dangle": [
      "error",
      "only-multiline",
    ],
    "comma-spacing": [
      "error",
      {
        "after": true,
        "before": false,
      },
    ],
    "curly": [
      "error",
      "multi-line",
      "consistent",
    ],
    "dot-notation": [
      "error",
    ],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": [
          "**/__tests__/**",
          "**/test/**",
        ],
        "optionalDependencies": false,
        "peerDependencies": true,
      },
    ],
    "import/no-unresolved": [
      "error",
    ],
    "import/order": [
      "warn",
      {
        "alphabetize": {
          "caseInsensitive": true,
          "order": "asc",
        },
        "groups": [
          "builtin",
          "external",
        ],
      },
    ],
    "key-spacing": [
      "error",
    ],
    "keyword-spacing": [
      "error",
    ],
    "max-len": [
      "error",
      {
        "code": 80,
        "ignoreComments": true,
        "ignoreRegExpLiterals": true,
        "ignoreStrings": true,
        "ignoreTemplateLiterals": true,
        "ignoreUrls": true,
      },
    ],
    "no-bitwise": [
      "error",
    ],
    "no-duplicate-imports": [
      "error",
    ],
    "no-multi-spaces": [
      "error",
      {
        "ignoreEOLComments": false,
      },
    ],
    "no-multiple-empty-lines": [
      "error",
    ],
    "no-return-await": [
      "off",
    ],
    "no-shadow": [
      "off",
    ],
    "no-trailing-spaces": [
      "error",
    ],
    "object-curly-newline": [
      "error",
      {
        "consistent": true,
        "multiline": true,
      },
    ],
    "object-curly-spacing": [
      "error",
      "always",
    ],
    "object-property-newline": [
      "error",
      {
        "allowAllPropertiesOnSameLine": true,
      },
    ],
    "quote-props": [
      "error",
      "consistent-as-needed",
    ],
    "quotes": [
      "error",
      "double",
      {
        "avoidEscape": true,
      },
    ],
    "semi": [
      "error",
      "always",
    ],
    "space-before-blocks": [
      "error",
    ],
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx",
      ],
    },
    "import/resolver": {
      "alias": {},
      "node": {
        "extensions": [
          ".js",
          ".jsx",
        ],
      },
      "typescript": undefined,
    },
  },
}
`;

exports[`NodeProject creates a generic NodeJS project with Eslint, Jest, and Typescript 1`] = `
{
  "clearMocks": true,
  "collectCoverage": true,
  "coverageDirectory": "coverage",
  "coveragePathIgnorePatterns": [
    "node_modules",
  ],
  "coverageReporters": [
    "json",
    "text",
  ],
  "globals": {
    "ts-jest": {
      "tsconfig": "tsconfig.json",
    },
  },
  "modulePathIgnorePatterns": [
    "node_modules",
  ],
  "preset": "ts-jest",
  "reporters": [
    "default",
    [
      "jest-junit",
      {
        "outputDirectory": "test-reports",
        "outputName": undefined,
      },
    ],
  ],
  "snapshotResolver": undefined,
  "testMatch": [
    "**/__tests__/?(*.)+(spec|test).[tj]s?(x)",
    "**/__tests__/**/?(*.)+(spec|test).[tj]s?(x)",
    "**/?(*.)+(spec|test).[tj]s?(x)",
  ],
  "testPathIgnorePatterns": [
    "node_modules",
  ],
  "watchPathIgnorePatterns": [
    "node_modules",
  ],
}
`;

exports[`NodeProject creates a generic NodeJS project with Eslint, Jest, and Typescript 2`] = `
{
  "env": {
    "jest": true,
    "node": true,
  },
  "extends": [
    "plugin:import/recommended",
    "plugin:import/typescript",
  ],
  "ignorePatterns": [
    "node_modules/",
    "*.generated.ts",
    "coverage",
    "*.js",
    "*.d.ts",
  ],
  "overrides": [
    {
      "files": [
        ".stackgenrc.ts",
      ],
      "rules": {
        "@typescript-eslint/no-require-imports": "off",
        "import/no-extraneous-dependencies": "off",
      },
    },
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2020,
    "project": "./tsconfig.json",
    "sourceType": "module",
  },
  "plugins": [
    "@typescript-eslint",
  ],
  "root": true,
  "rules": {
    "@typescript-eslint/member-delimiter-style": [
      "error",
    ],
    "@typescript-eslint/member-ordering": [
      "error",
      {
        "default": [
          "public-static-field",
          "public-static-method",
          "protected-static-field",
          "protected-static-method",
          "private-static-field",
          "private-static-method",
          "field",
          "constructor",
          "method",
        ],
      },
    ],
    "@typescript-eslint/no-floating-promises": [
      "error",
    ],
    "@typescript-eslint/no-require-imports": [
      "error",
    ],
    "@typescript-eslint/no-shadow": [
      "error",
    ],
    "@typescript-eslint/return-await": [
      "error",
    ],
    "array-bracket-newline": [
      "error",
      "consistent",
    ],
    "array-bracket-spacing": [
      "error",
      "never",
    ],
    "brace-style": [
      "error",
      "1tbs",
      {
        "allowSingleLine": true,
      },
    ],
    "comma-dangle": [
      "error",
      "only-multiline",
    ],
    "comma-spacing": [
      "error",
      {
        "after": true,
        "before": false,
      },
    ],
    "curly": [
      "error",
      "multi-line",
      "consistent",
    ],
    "dot-notation": [
      "error",
    ],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": [
          "**/__tests__/**",
          "**/test/**",
        ],
        "optionalDependencies": false,
        "peerDependencies": true,
      },
    ],
    "import/no-unresolved": [
      "error",
    ],
    "import/order": [
      "warn",
      {
        "alphabetize": {
          "caseInsensitive": true,
          "order": "asc",
        },
        "groups": [
          "builtin",
          "external",
        ],
      },
    ],
    "indent": [
      "off",
    ],
    "key-spacing": [
      "error",
    ],
    "keyword-spacing": [
      "error",
    ],
    "max-len": [
      "error",
      {
        "code": 80,
        "ignoreComments": true,
        "ignoreRegExpLiterals": true,
        "ignoreStrings": true,
        "ignoreTemplateLiterals": true,
        "ignoreUrls": true,
      },
    ],
    "no-bitwise": [
      "error",
    ],
    "no-duplicate-imports": [
      "error",
    ],
    "no-multi-spaces": [
      "error",
      {
        "ignoreEOLComments": false,
      },
    ],
    "no-multiple-empty-lines": [
      "error",
    ],
    "no-return-await": [
      "off",
    ],
    "no-shadow": [
      "off",
    ],
    "no-trailing-spaces": [
      "error",
    ],
    "object-curly-newline": [
      "error",
      {
        "consistent": true,
        "multiline": true,
      },
    ],
    "object-curly-spacing": [
      "error",
      "always",
    ],
    "object-property-newline": [
      "error",
      {
        "allowAllPropertiesOnSameLine": true,
      },
    ],
    "quote-props": [
      "error",
      "consistent-as-needed",
    ],
    "quotes": [
      "error",
      "double",
      {
        "avoidEscape": true,
      },
    ],
    "semi": [
      "error",
      "always",
    ],
    "space-before-blocks": [
      "error",
    ],
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx",
      ],
    },
    "import/resolver": {
      "alias": {},
      "node": {
        "extensions": [
          ".js",
          ".jsx",
          ".ts",
          ".tsx",
        ],
      },
      "typescript": {
        "alwaysTryTypes": true,
        "project": "tsconfig.json",
      },
    },
  },
}
`;

exports[`NodeProject creates a generic NodeJS project with Eslint, Jest, and Typescript 3`] = `
{
  "compilerOptions": {
    "alwaysStrict": true,
    "declaration": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "inlineSourceMap": true,
    "inlineSources": true,
    "lib": [
      "esnext",
    ],
    "module": "commonjs",
    "noEmit": false,
    "noEmitOnError": false,
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noUnusedLocals": true,
    "noUnusedParameters": false,
    "outDir": "build",
    "resolveJsonModule": true,
    "strict": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": true,
    "stripInternal": true,
    "target": "ES6",
  },
  "exclude": [
    "node_modules",
  ],
  "include": [
    "src/*.ts",
    "src/**/*.ts",
  ],
}
`;

exports[`NodeProject creates a generic NodeJS project with Jest 1`] = `
{
  "clearMocks": true,
  "collectCoverage": true,
  "coverageDirectory": "coverage",
  "coveragePathIgnorePatterns": [
    "node_modules",
  ],
  "coverageReporters": [
    "json",
    "text",
  ],
  "modulePathIgnorePatterns": [
    "node_modules",
  ],
  "reporters": [
    "default",
    [
      "jest-junit",
      {
        "outputDirectory": "test-reports",
        "outputName": undefined,
      },
    ],
  ],
  "snapshotResolver": undefined,
  "testMatch": [
    "**/__tests__/?(*.)+(spec|test).[tj]s?(x)",
    "**/__tests__/**/?(*.)+(spec|test).[tj]s?(x)",
    "**/?(*.)+(spec|test).[tj]s?(x)",
  ],
  "testPathIgnorePatterns": [
    "node_modules",
  ],
  "watchPathIgnorePatterns": [
    "node_modules",
  ],
}
`;

exports[`NodeProject creates a generic NodeJS project with Typescript 1`] = `
{
  "compilerOptions": {
    "alwaysStrict": true,
    "declaration": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "inlineSourceMap": true,
    "inlineSources": true,
    "lib": [
      "esnext",
    ],
    "module": "commonjs",
    "noEmit": false,
    "noEmitOnError": false,
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noUnusedLocals": true,
    "noUnusedParameters": false,
    "outDir": "build",
    "resolveJsonModule": true,
    "strict": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": true,
    "stripInternal": true,
    "target": "ES6",
  },
  "exclude": [
    "node_modules",
  ],
  "include": [
    "src/*.ts",
    "src/**/*.ts",
  ],
}
`;

exports[`NodeProject creates a generic NodeJS project with Yalc 1`] = `
{
  "bin": undefined,
  "bugs": undefined,
  "description": "MyDescription",
  "devDependencies": {
    "@stackgen/cli": "*",
    "@stackgen/core": "*",
  },
  "files": [
    "build/*.js",
    "build/**/*.js",
  ],
  "homepath": undefined,
  "keywords": undefined,
  "main": "build/index.js",
  "man": undefined,
  "name": "my-project",
  "private": true,
  "repository": {
    "type": "git",
    "url": "http://localhost",
  },
  "resolutions": {
    "a": "b",
  },
  "scripts": {
    "yalc": "npx yalc publish",
  },
  "types": undefined,
  "version": "0.0.1",
}
`;

exports[`NodeProject creates a generic NodeJS project with a license 1`] = `
{
  "bin": undefined,
  "bugs": undefined,
  "description": "MyDescription",
  "devDependencies": {
    "@stackgen/cli": "*",
    "@stackgen/core": "*",
  },
  "files": [
    "build/*.js",
    "build/**/*.js",
  ],
  "homepath": undefined,
  "keywords": undefined,
  "license": "Apache-2.0",
  "main": "build/index.js",
  "man": undefined,
  "name": "my-project",
  "private": true,
  "repository": {
    "type": "git",
    "url": "http://localhost",
  },
  "resolutions": {
    "a": "b",
  },
  "scripts": {
    "yalc": "npx yalc publish",
  },
  "types": undefined,
  "version": "0.0.1",
}
`;

exports[`NodeProject creates a generic NodeJS project with overloaded packageName 1`] = `
{
  "bin": undefined,
  "bugs": undefined,
  "description": "MyDescription",
  "devDependencies": {
    "@stackgen/cli": "*",
    "@stackgen/core": "*",
  },
  "files": [
    "build/*.js",
    "build/**/*.js",
  ],
  "homepath": undefined,
  "keywords": undefined,
  "main": "build/index.js",
  "man": undefined,
  "name": "@example/test",
  "private": true,
  "repository": {
    "type": "git",
    "url": "http://localhost",
  },
  "resolutions": {
    "a": "b",
  },
  "scripts": {
    "yalc": "npx yalc publish",
  },
  "types": undefined,
  "version": "0.0.1",
}
`;
